### ISSUE TITLE
JSON conversion does not keep some of original data types.

### ISSUE THREAD

**Issue by SnowMarble** [ramiel/mikro-orm-cache-adapter-redis#6]
Currently, `JSON.stringify` and `JSON.parse` are used to stringify objects. But, some data types like typed arrays, cannot be converted to their original type .

## Reproduction 

Let's assume we have this POJO data.
```ts
{
  id: '7cef963a-895e-4e1f-b93e-9f2085e38e87',
  createdAt: new Date('2024-04-04T21:06:01.463Z'),
  updatedAt: new Date('2024-04-04T21:06:01.463Z'),
  key: Buffer.from([
    65, 65, 207, 252, 215, 0, 245, 135, 1, 7, 165, 188, 172, 33, 19, 215,
  ]),
  blob: new Uint8ClampedArray([1, 2, 3, 4]),
}
```

This is stringified to:
```text
{
  "id": "7cef963a-895e-4e1f-b93e-9f2085e38e87",
  "createdAt": "2024-04-04T21:06:01.463Z",
  "updatedAt": "2024-04-04T21:06:01.463Z",
  "key": {
    "type": "Buffer",
    "data": [65,65,207,252,215,0,245,135,1,7,165,188,172,33,19,215]
  },
  "blob": {
    "0": 1,
    "1": 2,
    "2": 3,
    "3": 4
  }
}
```

And, it's parsed to:
```ts
{
  id: "7cef963a-895e-4e1f-b93e-9f2085e38e87",
  createdAt: "2024-04-04T21:06:01.463Z",
  updatedAt: "2024-04-04T21:06:01.463Z",
  key: {
    type: "Buffer",
    data: [
      65, 65, 207, 252, 215, 0, 245, 135, 1, 7, 165, 188, 172, 33, 19, 215
    ],
  },
  blob: {
    "0": 1,
    "1": 2,
    "2": 3,
    "3": 4,
  },
}
```

Typed array and  Buffer are not parsed to their original type.

We need to keep their types after parsing.

## List of primitives to keep their original type

here's a list of primitives to keep their original type I found so far:

- typed array (<https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Typed_arrays#typed_array_views>)
- Buffer (from Node, Bun ...)

## Other primitives

I've just found this issue, and I'm checking if other data types like Blob, Symbol, or Date are to be considered. 

Please let me know if any of you found other data types to manage.

**Comment by ramiel** [comment]
Good findings. Any PR welcome of course

**Comment by SnowMarble** [comment]
One of solution is to use default serializer and deserializer from v8 api. 

```ts
import assert from 'node:assert'
import { deserialize, serialize } from 'node:v8'

const obj = {
  str: 'string',
  num: 0,
  obj: { foo: 'foo' },
  arr: [1, 2, 3],
  bool: true,
  nil: null,
  undef: undefined,
  inf: Infinity,
  date: new Date('Thu, 28 Apr 2016 22:02:17 GMT'),
  map: new Map([['hello', 'world']]),
  set: new Set([123, 456]),
  buffer: Buffer.from([1, 2, 3]),
  typedArray: new Uint8Array([1, 2, 3]),
  typedArray2: new Uint8ClampedArray([1, 2, 3]),
  re: /([^\s]+)/g,
  big: BigInt(10),
}

console.log('original object')
console.log(obj)

const s = serialize(obj)

console.log('serialized')
console.log(s.toString('base64'))

const d = deserialize(s)

console.log('deserialized')
console.log(d)

assert.deepEqual(obj, d)
```

This can serialize most of data types, but fails when the object contains function or URL instance.

But we can just handle these special cases.

Also, the serialized data is binary, we might need to use `JSON.stringify` for debug.

What do you think of it?

**Comment by SnowMarble** [comment]
@ramiel  I tried to find the best data serializer for Node and Bun, I concluded the v8 `serialize`/`deserialize` performs the best.

While JSON serialization doesn't consistently reproduce the exact same data after deserialization, v8 does so reliably when both serializing and deserializing are successful.

Also, since Redis can handle binary data, we can directly store the output. We can still generate human-readable serialization results using `JSON.stringify` for debugging purposes. If you agree with my findings, I can work on it right away. 😄

**Comment by ramiel** [comment]
I'm thinking a bit more about this issue. I'm realizing we don't need to serialize every possible javascript type, but only what can be saved/loaded from mikro-orm and so what can be saved into the database. So, my first question is how did you managed to have a Buffer from mikro-orm?

**Comment by SnowMarble** [comment]
Yes, I'm using Blob Type provided by mikro-orm. But I didn't made this issue because the Blob Type doesn't have an appropriate `toJSON` method.

The data received by the cache adapter is an entity that has been processed using `toPOJO` (`wrap(entity).toPOJO()`).
If mikro-orm only allowed properties stored directly into the database, we would simply need to handle a few edge cases like Buffer and Typed Array.

However, mikro-orm provides virtual properties that are not stored in the database but can be accessed at the programming level. This means an entity can contain any primitive type - regex, function, Map, and so forth. (Since we don't store the virtual properties, custom types are unnecessary.)

So, in theory, every JavaScript type can be loaded from mikro-orm.

The virtual properties remain after the `toPOJO` processing. Consequently, the cache adapter must be capable of managing all primitives.

**Comment by ramiel** [comment]
In this case, please go on. Thank you!

**Comment by Stene3** [comment]
@ramiel Any chance it will get reviewed and will be merged to the stable release soon? This is very needed even for Date types and v8 serializer solves this. Thanks

**Comment by SnowMarble** [comment]
@Stene3 I actually made my own version of the cache adapter. You can check it out here:
https://www.npmjs.com/package/@javien/mikro-orm-redis-cache-adapter

**Comment by ramiel** [comment]
Hey guys, totally forgot about this. Let me merge and release

**Comment by ramiel** [comment]
Merged and released version 4.0.0
